# ISSAAC-seq preprocessing pipeline v2
ISSAAC-seq is a single-cell multiomic method for simultaneous profiling of chromatin accessibility and gene expression from the same cell. This repository contains an updated version of the preprocessing pipeline compared to the [original publication](https://www.nature.com/articles/s41592-022-01601-4) back in 2022.

## Raw Sequencing Data

The 10x Genomics workflow generates data in the same format as the original publication, and you can find the raw data (fastq files) from ArrayExpress under the accession number [E-MTAB-11264](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-11264/). For the plate-based workflow, we now put the well barcode to the `i5` position so that it has the same library format as the 10x Genomics workflow, and the raw data generated with PBMCs from a healthy individual can be found under the accession number [E-MTAB-15131](https://www.ebi.ac.uk/biostudies/ArrayExpress/studies/E-MTAB-15131). The Bio-Rad workflow has different library format, and the raw data can be found under the same accession number [E-MTAB-15131](https://www.ebi.ac.uk/biostudies/ArrayExpress/studies/E-MTAB-15131).

## Analysing your own data

The pipeline chains a number of commonly used programmes together, which means we need to make sure all of them are properly set up. Luckily, we only need to do this once at the very beginning of our project.

### Step 1: Software Installation

The following programmes and packages are needed to run the full pipeline. Make sure they are executable and in our `$PATH` so that they can be invoked directly without having to specify the absolute path.

- [cutadapt](https://cutadapt.readthedocs.io) (v4.5)
- [chromap](https://github.com/haowenz/chromap) (0.2.1-r369)
- [MACS2](https://pypi.org/project/MACS2/) (v2.2.7.1)
- [STARsolo](https://github.com/alexdobin/STAR) (v2.7.11b)
- [samtools](https://www.htslib.org) (v1.9)
- [bedtools](https://bedtools.readthedocs.io/en/latest/) (v2.30.0)
- [seqtk](https://github.com/lh3/seqtk) (v1.3-r106)
- [bgzip](https://www.htslib.org/doc/bgzip.html)
- [tabix](https://www.htslib.org/doc/tabix.html)
- The following pre-compiled binaries from [UCSC utilities](http://hgdownload.soe.ucsc.edu/admin/exe/):
  - [addCols](http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/addCols)
  - [bedClip](http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bedClip)
  - [bedGraphToBigWig](http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bedGraphToBigWig)
  - [calc](http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/calc)
  - [faSize](http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/faSize)

To convert the `bdg` file generated by MACS2 to the `bw` format for read pileup visualisation on the UCSC genome browser, we also need the [bdg2bw](https://gist.github.com/taoliu/2469050) script. In this pipeline, a modified version of the script is used, and the script is provided in the `misc_files` folder. We need to move it to our `$PATH` and make sure it is executable.

The pipeline also uses `python3` and the following python packages, all of which can be installed via [miniconda](https://www.anaconda.com/docs/getting-started/miniconda/main) or the like:

```
scipy v1.11.4
numpy v1.26.0
pandas v2.1.2
matplotlib v3.9.1
```

### Step 2: Preparation of genome files

Once we have finished installing all the programmes and packages, we need to prepare some files related to the genome of our interest. First, we generate star index for the genome:

```bash
STAR \
    --runThreadN <num_threads> \
    --runMode genomeGenerate \
    --genomeDir <output_dir_that_will_host_the_output_files> \
    --genomeFastaFiles <path_to_the_genome_fasta_file> \
    --sjdbGTFfile <path_to_the_gene_annotation_gtf_file>
```

Then we generate the chromap index for the genome:

```bash
chromap \
    -i \
    -t <num_threads> \
    -r <path_to_the_genome_fasta_file> \
    -o <output_index_file>
```

Next, we need to provide a whitelist for the cell barcodes, which are the index sequences on the gel beads (droplet) or the `i5` position of the primers (plate). Since we are using the 10x Single Cell ATAC kit, we need the 16-bp cell barcodes from the cellranger-atac software. I have prepared this file together with our plate barcode related files in the `whitelist` directory of this repository, like this:

```
whitelist/
├── 737K-cratac-v1_rc.txt                            # 10x single cell atac whitelist
├── ISSAACv2_plate_whitelist.tsv                     # whitelist for plate-based method (16 bp)
├── ISSAACv2_plate_whitelist_first8.tsv              # whitelist for plate-based method (first 8 bp, useful when only 8 cycles are sequenced)
├── ISSAACv2_plate_whitelist_with_wellid.tsv         # tab-delimited file indicating the well position of each barcode (16 bp), used to draw qc
└── ISSAACv2_plate_whitelist_first8_with_wellid.tsv  # tab-delimited file indicating the well position of each barcode (the first 8 bp), used to draw qc
```

Then, we need to prepare the blacklist regions for the genome, which is the problematic regions with abnormally high signals in many sequencing experiments. The details can be found in the [ENCODE blacklist paper](https://www.nature.com/articles/s41598-019-45839-z) and the corresponding [GitHub repository](https://github.com/Boyle-Lab/Blacklist). I have provided the blacklist regions of human (`hg38`) and mouse (`mm10`) under the `misc_files` directory of this repository.

Another file that is needed in various steps within the pipeline is the chromosomal sizes file. It is a tab-delimited file with two columns. The first column is the name of the chromosome, and the second column is the size in bp of the chromosome. We can generate the file of the genome of our interest using the binary [faSize](http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/faSize) and the genome fasta file like this:

```
faSize -detailed <genome_fasta> | \
    sort -k1,1 > chrom.sizes.sorted.tsv
```

I have provided the files for human (`hg38`) and mouse (`mouse`) under the `misc_files` directory of this repository.

### Step 3: Preparation of your own data

Put the path to the corresponding FASTQ files in `config.json`. To run the pipeline using all available cores, simply run:

```
snakemake --cores
```

## Softwares/Packages prerequisite:

1. __General programs__
  - Python3
  - R v4
2. __Python packages__
  - numpy (v1.19.5)
  - scipy (v1.5.3)
  - pandas (v1.1.5)
  - seaborn (v0.11.0.rc0)
  - scanpy (v1.9.0)
  - scikit-learn (v0.22.1)
3. __R packages__
  - Matrix (v1.3-3)
  - Signac (v1.6.0)
  - Seurat (v4.1.0)
  - monocle (v2.20.0)
  - GenomeInfoDb (v1.30.1)
  - EnsDb.Mmusculus.v79 (v2.99.0)
  - EnsDb.Hsapiens.v86 (v2.99.0)
  - ggplot2 (v3.35)
  - patchwork (v1.1.1)
  - dplyr (v1.0.8)
  - stringr (v1.4.0.9000)
  - mgsub (v1.7.3)
  - stringi (v1.7.6)

## Contact

Xi Chen  
chenx9@sustech.edu.cn
